### PR TITLE
Query key values

### PR DESCRIPTION
**Pull request by glynos**
Applied a solution to this issue:

https://github.com/cpp-netlib/uri/issues/96

### CODE REVIEW

**Comment by deanberris** [comment]
LGTM

File: test/uri_test.cpp [hunk 1]
```diff
@@ -820,3 +820,67 @@ TEST(uri_test, non_opaque_path_has_double_slash) {
   EXPECT_EQ("/path/to/something/", instance.path());
   EXPECT_FALSE(instance.is_opaque());
 }
+
+TEST(uri_test, query_iterator_with_no_query) {
+  network::uri instance("http://example.com/");
+  ASSERT_FALSE(instance.has_query());
```
Comment (georgi-d) [inline test/uri_test.cpp]: I would add an additional test:
```
EXPECT_EQ(instance.query_begin(), instance.query_end());
```

Here to validate that there are no query {k, v} through the interator interface

File: test/uri_test.cpp [hunk 2]
```diff
@@ -820,3 +820,67 @@ TEST(uri_test, non_opaque_path_has_double_slash) {
...
+  EXPECT_EQ(instance.query_begin(), instance.query_end());
+}
+
+TEST(uri_test, query_iterator_with_single_kvp) {
+  network::uri instance("http://example.com/?a=b");
+  ASSERT_TRUE(instance.has_query());
+  ASSERT_NE(instance.query_begin(), instance.query_end());
+  auto kvp = *instance.query_begin();
+  EXPECT_EQ("a", kvp.first);
+  EXPECT_EQ("b", kvp.second);
```
Comment (georgi-d) [inline test/uri_test.cpp]: I would add additional check after this line:

```
++query_it;
EXPECT_EQ(query_it, instance.query_end());
```

To validate that query_end() is properly reached.

File: test/uri_test.cpp [hunk 3]
```diff
@@ -820,3 +820,67 @@ TEST(uri_test, non_opaque_path_has_double_slash) {
...
+  EXPECT_EQ("c", query_it->first);
+  EXPECT_EQ("d", query_it->second);
+}
+
+TEST(uri_test, query_iterator_with_key_and_no_value) {
+  network::uri instance("http://example.com/?query");
+  ASSERT_TRUE(instance.has_query());
+  auto query_it = instance.query_begin();
+  EXPECT_EQ("query", query_it->first);
+  EXPECT_EQ("", query_it->second);
```
Comment (georgi-d) [inline test/uri_test.cpp]: I would add additional check after this line:
```
++query_it;
EXPECT_EQ(query_it, instance.query_end());
```

To validate that query_end() is properly reached.

File: test/uri_test.cpp [hunk 4]
```diff
@@ -820,3 +820,67 @@ TEST(uri_test, non_opaque_path_has_double_slash) {
...
+TEST(uri_test, query_iterator_with_fragment) {
+  network::uri instance("http://example.com/?a=b;c=d#fragment");
+  ASSERT_TRUE(instance.has_query());
+  ASSERT_NE(instance.query_begin(), instance.query_end());
+  auto query_it = instance.query_begin();
+  EXPECT_EQ("a", query_it->first);
+  EXPECT_EQ("b", query_it->second);
+  ++query_it;
+  EXPECT_EQ("c", query_it->first);
+  EXPECT_EQ("d", query_it->second);
```
Comment (georgi-d) [inline test/uri_test.cpp]: I would add additional check after this line:
```
++query_it;
EXPECT_EQ(query_it, instance.query_end());
```

To validate that query_end() is properly reached.

**Comment by glynos** [comment]
@georgi-d: I only saw your suggestions after the PR was merged. Today I added them to the test set and ... the query_iterator was all kinds of broken. I made a new PR (#102). Can you confirm that this works for you? Thanks.

**Pull request merged**